### PR TITLE
Add country and county dropdowns

### PR DESCRIPTION
**Pull request by MorayM**
Switches all uses of countries and counties to use drop downs instead of free text. Determining whether an address or country of birth is in the UK is driven from the dropdown instead of being a separate field. It's not fully tested yet (!) so caveat mergor.

image

image

### CODE REVIEW

File: GLAA.Domain/GLAAContext.cs [hunk 1]
```diff
@@ -46,18 +48,18 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
             }
 
             // many-to-many
-            modelBuilder.Entity<LicenceCountry>()
-                .HasKey(x => new { x.LicenceId, x.CountryId });
+            modelBuilder.Entity<LicenceWorkerCountry>()
```
Comment (dougajmcdonald) [inline GLAA.Domain/GLAAContext.cs:51]: Is there no longer a need to have a country assigned to a licence (the operating country for example)? or is that handled elsewhere?

File: GLAA.Domain/Models/Address.cs [hunk 2]
```diff
@@ -14,12 +15,14 @@ public class Address : IId
...
-        public string County { get; set; }
-
         public string Postcode { get; set; }
 
-        public string Country { get; set; }
+        public int? CountryId { get; set; }
+        [ForeignKey(nameof(CountryId))]
+        public virtual Country Country { get; set; }
 
-        public bool NonUK { get; set; }
```
Comment (dougajmcdonald) [inline GLAA.Domain/Models/Address.cs:23]: With the removal of the NonUK flag, does this force the user to enter data under the guise of mandatory which may then be set to non mandatory when they select the country? I assume this is all agreed wireframed etc? Reason I mention is is that DD has quite vocal on having 'Non UK' as a flag before filling in the address info.

File: GLAA.Services/AccountCreation/AccountCreationViewModelBuilder.cs [hunk 3]
```diff
@@ -9,22 +9,23 @@ public class AccountCreationViewModelBuilder : IAccountCreationViewModelBuilder
     {
         private readonly IMapper mapper;
         private readonly UserManager<GLAAUser> userManager;
+        private readonly IReferenceDataProvider referenceDataProvider;
 
-        public AccountCreationViewModelBuilder(IMapper mp, UserManager<GLAAUser> um)
+        public AccountCreationViewModelBuilder(IMapper mp, UserManager<GLAAUser> um, IReferenceDataProvider rdp)
```
Comment (dougajmcdonald) [inline GLAA.Services/AccountCreation/AccountCreationViewModelBuilder.cs:14]: Cool, this is how I'd imagined this guy working

File: GLAA.ViewModels/LicenceApplication/AlternativeBusinessRepresentativeViewModel.cs [hunk 4]
```diff
@@ -1,50 +1,12 @@
...
 {
-    public class AlternativeBusinessRepresentativeViewModel : PersonViewModel, IValidatable
+    public class AlternativeBusinessRepresentativeViewModel : PersonViewModel
     {
         public AlternativeBusinessRepresentativeViewModel()
         {
             RightToWorkViewModel = new RightToWorkViewModel();
         }
 
-        public void Validate()
```
Comment (dougajmcdonald) [inline GLAA.ViewModels/LicenceApplication/AlternativeBusinessRepresentativeViewModel.cs:14]: Nice

File: GLAA.ViewModels/LicenceApplication/DirectorOrPartnerViewModel.cs [hunk 5]
```diff
@@ -15,41 +15,7 @@ public DirectorOrPartnerViewModel()
         public IsPreviousPrincipalAuthorityViewModel IsPreviousPrincipalAuthority { get; set; }
 
         public int? PrincipalAuthorityId { get; set; }
-
-        public void Validate()
```
Comment (dougajmcdonald) [inline GLAA.ViewModels/LicenceApplication/DirectorOrPartnerViewModel.cs:19]: Number 2! Kill it! :)

File: GLAA.Web/Controllers/OrganisationDetailsController.cs [hunk 6]
```diff
@@ -24,12 +24,13 @@ public class OrganisationDetailsController : LicenceApplicationBaseController
 
         [HttpGet]
         [ImportModelState]
+        //public IActionResult Part(int id, bool? back = false, bool? nonuk = null, bool? manual = null)
```
Comment (dougajmcdonald) [inline GLAA.Web/Controllers/OrganisationDetailsController.cs]: Cull?

File: GLAA.Web/Helpers/HtmlHelpers.cs [hunk 7]
```diff
@@ -72,6 +72,13 @@ public static IHtmlContent InternalStatusFor<TModel>(this IHtmlHelper<TModel> ht
             return BuildFormGroupForControl(html, expression, html.EditorFor(expression, "_NullableTimeSpan"));
         }
 
+        public static IHtmlContent DropDownFormGroupFor<TModel, TResult>(this IHtmlHelper<TModel> html,
+            Expression<Func<TModel, TResult>> expression, IEnumerable<SelectListItem> items)
+        {
+            var ddl = html.DropDownListFor(expression, items.OrderBy(i => i.Text).ToArray(), "", new { @class = "form-control"});
```
Comment (dougajmcdonald) [inline GLAA.Web/Helpers/HtmlHelpers.cs:78]: Can we ensure the dropdowns are styled like the GDS ones pls? https://govuk-elements.herokuapp.com/form-elements/#form-select-boxes

File: GLAA.Web/Startup.cs [hunk 8]
```diff
@@ -125,6 +125,7 @@ public void ConfigureServices(IServiceCollection services)
             services.AddTransient<IFormDefinition, LicenceApplicationFormDefinition>();
             services.AddTransient<IFieldConfiguration, FieldConfiguration>();
             services.AddTransient<IConstantService, ConstantService>();
+            services.AddTransient<IReferenceDataProvider, ReferenceDataProvider>();
```
Comment (dougajmcdonald) [inline GLAA.Web/Startup.cs]: If we're not changing country and county at runtime, can this be a singleton and get the data on instantiation? just thinking we only need to get the data once for the app lifetime really don't we?

**Review by dougajmcdonald — APPROVED** [review]
Looks good, couple of minor comments, shout if it makes no sense!

File: GLAA.Domain/GLAAContext.cs (hunk 1 above)
Comment (MorayM) [reply GLAA.Domain/GLAAContext.cs:51]: It's a straight rename of the class, the functionality is still there 😄

File: GLAA.Domain/Models/Address.cs (hunk 2 above)
Comment (MorayM) [reply GLAA.Domain/Models/Address.cs:23]: It's removed in Adam's wireframes, which I assume DD has seen? There is still a concept of a UK/Non UK address although the logic is driven by the country dropdown. The flow is slightly different as a user can say that they have an address outside the UK and still enter "UK - England" but the validation (when it's finished) will force them to fill in the appropriate fields for what they've selected.

File: GLAA.Web/Controllers/OrganisationDetailsController.cs (hunk 6 above)
Comment (MorayM) [reply GLAA.Web/Controllers/OrganisationDetailsController.cs]: WIP!

File: GLAA.Web/Helpers/HtmlHelpers.cs (hunk 7 above)
Comment (MorayM) [reply GLAA.Web/Helpers/HtmlHelpers.cs:78]: They are, `.form-control` handles it.

File: GLAA.Web/Startup.cs (hunk 8 above)
Comment (MorayM) [reply GLAA.Web/Startup.cs]: Good spot!

**Pull request merged**